mavros_extras/src/plugins/landing_target.cpp

```diff
@@ -0,0 +1,198 @@
+											<< " -> " << child_frame_id);
+			tf_start("LandingTargetTF", &LandingTargetPlugin::send_landing_target);
+		}
+
+		land_target_sub = sp_nh.subscribe("target", 10, &LandingTargetPlugin::land_target_cb, this);
+		local_position_sub = sp_nh.subscribe("/mavros/position/local", 10, &LandingTargetPlugin::local_pos_cb, this);
+	}
+
+	const message_map get_rx_handlers() {
+		return {
+			       MESSAGE_HANDLER(MAVLINK_MSG_ID_LANDING_TARGET, &LandingTargetPlugin::handle_landing_target)
+		};
+	}
+
+private:
+	friend class TFListenerMixin;
+	ros::NodeHandle sp_nh;
+	UAS *uas;
+
+	bool send_tf;
+	bool listen_tf;
+	double tf_rate;
+	ros::Time last_transform_stamp;
+	tf::TransformBroadcaster tf_broadcaster;
+
+	tf::Transform transform;
+	ros::Time lp_time;
+
+	std::string frame_id;
+	std::string child_frame_id;
+
+	ros::Publisher land_target_pub;
+	ros::Subscriber land_target_sub;
+	ros::Subscriber local_position_sub;
+
+	/* -*- low-level send -*- */
+
+	void landing_target(uint8_t target_num,
+			uint8_t frame,
+			float angle_x,
+			float angle_y,
+			float distance) {
+		mavlink_message_t msg;
+		mavlink_msg_landing_target_pack_chan(UAS_PACK_CHAN(uas), &msg,
+				target_num,
+				frame,
+				angle_x,
+				angle_y,
+				distance);
+		UAS_FCU(uas)->send_message(&msg);
+	}
+
+	/* -*- mid-level helpers -*- */
+
+	/*	AUXILIAR MATH:
+	*	d = sqrt(x^2 + y^2 + z^2)
+	*	x = sqrt((d^2-z^2) * cos^2(ax))
+	*	y = x*tan(ax)
+	*	z = d^2 - x^2 - y^2;
+	*/
+
+	/**
+	 * Send landing target transform to FCU
+	 */
+	void send_landing_target(const tf::Transform &transf, const ros::Time &stamp) {
+		// origin position in ROS ENU frame
+		tf::Vector3 pos = transf.getOrigin();
+
+		float ang_x = atan(pos.y()/pos.x());
+		float ang_y = boost::math::constants::pi<double>()/2 - atan(pos.y()/pos.x());
+		float distance = sqrt(pow(pos.x(),2) + pow(pos.y(),2) + pow(pos.z(),2));
+
+		if (last_transform_stamp == stamp) {
+			ROS_DEBUG_THROTTLE_NAMED(10, "landing_target", "Target: Same transform as last one, dropped.");
+			return;
+		}
+		last_transform_stamp = stamp;
+
+		landing_target( 0, 		// TODO: update number depending on received frame_id
+				1, 		// in NED; should user choose or do we define it?
+				ang_x, -ang_y,	// which may mean this angles should be adapted to frame
+				distance);	// TODO: add MAV_FRAME enum to uas
+	}
+
+	void handle_landing_target(const mavlink_message_t *msg, uint8_t sysid, uint8_t compid) {
+		mavlink_landing_target_t land_target;
+		mavlink_msg_landing_target_decode(msg, &land_target);
+
+		ROS_DEBUG_THROTTLE_NAMED(10, "land_target", "Landing target: "
+				"frame: %d angle offset:(X: %1.3frad, Y: %1.3frad) distance: %1.3fm",
+				land_target.frame, land_target.angle_x, land_target.angle_y, land_target.distance);
+
+		tf::Vector3 target;
+		tf::Vector3 local = transform.getOrigin();
+
+		target.setX(sqrt((pow(land_target.distance,2) - pow(local.z(),2)) * pow(cos(land_target.angle_x),2)));
```

Comment:
If I get this right, this representation is suppose to be a reference to angular offsets in spherical coordinates, which means that `angle_x` should be `theta` and `angle_y` should be `phi` and that way we have `position.x = r*sin(phi)*cos(theta)`, `position.y = r*sin(phi)*sin(theta)` and `position.z = r*cos(theta)`, where `theta` gets values from 0 to 2PI and theta from 0 to PI and `r` is the distance.
The current angle names led me to think they were complementary and that we were lacking an vertical offset. If this was present as spherical coordinates it would make much more sense.